### PR TITLE
Initial cost and income mapping based on RAW AnM AAR data.

### PR DESCRIPTION
**Pull request by colinbull**
### Context
Include relevant motivation and context.

### Change proposed in this pull request
Include a summary of the change.

### Guidance to review 
Include any useful information needed to review this change. Include any dependencies that are required for this change. 

### Checklist
- [x] Work items have been linked [(use AB#)](https://learn.microsoft.com/en-us/azure/devops/boards/github/link-to-from-github?view=azure-devops#use-ab-to-link-from-github-to-azure-boards-work-items)
- [x] Your code builds clean without any errors or warnings
- [x] You have run all unit/integration tests and they pass
- [x] Your branch has been rebased onto main
- [x] You have tested by running locally

### CODE REVIEW

File: data-pipeline/src/pipeline/input_schemas.py [hunk 1]
```diff
@@ -262,146 +262,122 @@
...
+    "BNCH11162 (Academies)": "float",
+    "BNCH11163 (Non- Government)": "float",
+    "BNCH11123-BAI011-A (Academies - Income)": "float",
+    "BNCH11201 (Income from facilities and services)": "float",
+    "BNCH11202 (Income from catering)": "float",
+    "BNCH11203 (Receipts from supply teacher insurance claims)": "float",
+    "BNCH11300T (Voluntary income)": "float",
+    "BNCH11204 (Other income - revenue)": "float",
+    "BNCH11205 (Other Income from facilities and services)": "float",
+    "BNCH11400T (Investment income)": "float",
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/input_schemas.py:289]: Do we know what the difference is between `BNCH11401 (Investment income)` and `BNCH11400T (Investment income)`?

File: data-pipeline/src/pipeline/input_schemas.py [hunk 2]
```diff
@@ -262,146 +262,122 @@
...
+    "BNCH21801 (Interest charges for Loan and bank)": "float",
+    "BNCH21705 (Other insurance premiums)": "float",
+    "BNCH21802 (PFI Charges)": "float",
+    "BNCH21404 (Rent and rates)": "float",
+    "BNCH21501 (Special facilities)": "float",
+    "BNCH21202 (Staff development and training)": "float",
+    "BNCH21203 (Staff-related insurance)": "float",
+    "BNCH21204 (Supply teacher insurance)": "float",
+    "BNCH21401 (Cleaning and caretaking)": "float",
+    "BNCH21406 (Other occupation costs)": "float",
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/input_schemas.py:316]: Any difference between `BNCH21406 (Other occupation costs)` and `BNCH21400T (Other occupancy costs)`?

File: data-pipeline/src/pipeline/input_schemas.py [hunk 3]
```diff
@@ -262,146 +262,122 @@
...
+    "BNCH21105 (Premises staff)": "float",
+    "BNCH21101 (Teaching staff)": "float",
+    "BNCH21102 (Supply teaching staff - extra note in guidance)": "float",
+    "BNCH21103 (Education support staff)": "float",
+    "BNCH21604 (Educational Consultancy)": "float",
+    "BNCH21606 (Agency supply teaching staff)": "float",
+    "BNCH21403 (Energy)": "float",
+    "BNCH21402 (Water and sewerage)": "float",
 }
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/input_schemas.py:325]: Do you still need to map `BNCHBAI061 (Coronavirus Govt Funding)` for the historic data? Or are we leaving that out following our conversation earlier?

File: data-pipeline/src/pipeline/input_schemas.py [hunk 4]
```diff
@@ -262,146 +262,122 @@
...
-    "Date joined or opened if in period": "string",
-    "Date left or closed if in period": "string",
+    "ACADEMYUPIN": "Int64",
+    "Company_Number": "string",
+    "Date joined or opened if in period:": "string",
+    "Date left or closed if in period:": "string",
+    "BNCH11110T (EFA Revenue Grants)": "float",
+    "BNCH11131 (DfE Family Revenue Grants)": "float",
+    "BNCH11141 (SEN)": "float",
+    "BNCH11142 (Other Revenue)": "float",
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/input_schemas.py:277]: What is the difference between this and `BNCH11160T (Other Revenue Income)`?

File: data-pipeline/src/pipeline/input_schemas.py [hunk 5]
```diff
@@ -262,146 +262,122 @@
...
-    "Staff Total ": "float",
-    "Maintenance & Improvement": "float",
-    "Premises": "float",
-    "Catering Exp": "float",
-    "Occupation": "float",
-    "Supplies and Services": "float",
-    "Educational Supplies": "float",
-    "Brought in Professional Services": "float",
-    "Total Expenditure": "float",
+    "Lead_UPIN": "Int64",
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/input_schemas.py:329]: Do we also not want to retain the company number?

File: data-pipeline/src/pipeline/input_schemas.py [hunk 6]
```diff
@@ -262,146 +262,122 @@
...
+    "BNCH11151 (Other Government Revenue Grants)": "float",
+    "BNCH11161 (Government source (non-grant))": "float",
+    "BNCH11162 (Academies)": "float",
+    "BNCH11163 (Non- Government)": "float",
+    "BNCH11123-BTI011-A (MAT Central services - Income)": "float",
+    "BNCH11201 (Income from facilities and services)": "float",
+    "BNCH11202 (Income from catering)": "float",
+    "BNCH11203 (Receipts from supply teacher insurance claims)": "float",
+    "BNCH11300T (Voluntary income)": "float",
+    "BNCH11204 (Other income - revenue)": "float",
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/input_schemas.py:343]: Any difference between these two? 
`BNCH11160T (Other Revenue Income)` and `BNCH11204 (Other income - revenue)`

File: data-pipeline/src/pipeline/input_schemas.py [hunk 7]
```diff
@@ -262,146 +262,122 @@
...
+    "BNCH11205 (Other Income from facilities and services)": "float",
+    "BNCH11400T (Investment income)": "float",
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/input_schemas.py:345]: Any difference between this and `BNCH11401 (Investment income)`?

File: data-pipeline/src/pipeline/mappings.py [hunk 8]
```diff
@@ -263,3 +259,49 @@ def map_cost_series(category_name, df, basis):
...
+        match f"{la}{estab}":
+            case "3066900":
+                return "Outer"
+            case "9332064":
+                return "Neither"
+            case "8946900":
+                return "Neither"
+            case "3906900":
+                return "Neither"
+            case _:
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/mappings.py:301]: What is the reason for focusing on these specific LAEstabs in particular?

File: data-pipeline/tests/unit/pre_processing/test_aar.py [hunk 9]
```diff
@@ -6,114 +6,98 @@
 
 def test_aar_data_has_correct_output_columns(prepared_aar_data: pd.DataFrame):
     assert list(prepared_aar_data.columns) == [
+        "LA",
+        "Estab",
         "Academy UPIN",
-        "In year balance",
-        "PFI School",
-        "Trust UPIN",
```
Comment (Chris-Lorch) [inline data-pipeline/tests/unit/pre_processing/test_aar.py:12]: Wont we need the TrustUPIN for the trust table, which we use to link it to the BFR data?

File: data-pipeline/tests/unit/pre_processing/test_mappings.py [hunk 10]
```diff
@@ -297,13 +297,13 @@ def test_map_school_website(url, expected):
 @pytest.mark.parametrize(
     "pfi,expected",
     [
-        ("part of PFI", "PFI School"),
-        ("not PFI", "Non-PFI school"),
-        ("", "Non-PFI school"),
+        (1, "PFI School"),
+        (0, "Non-PFI school"),
+        (-1, "PFI School"),
```
Comment (Chris-Lorch) [inline data-pipeline/tests/unit/pre_processing/test_mappings.py:302]: might be a silly question, but if we're just checking the integers, why not use the absolute and remove line 302?

File: data-pipeline/tests/unit/pre_processing/test_aar.py [hunk 11]
```diff
@@ -6,114 +6,98 @@
...
+        "BNCH11110T (EFA Revenue Grants)",
+        "BNCH11131 (DfE Family Revenue Grants)",
+        "Income_Targeted grants",
         "Income_Pre Post 16",
         "Income_Other grants",
         "Income_Government source",
         "Income_Academies",
         "Income_Non government",
-        "Income_Facilities and services",
+        "BNCH11123-BAI011-A (Academies - Income)",
```
Comment (Chris-Lorch) [inline data-pipeline/tests/unit/pre_processing/test_aar.py:22]: Should we be passing the raw data references to our outputs? Shouldn't they be mapped to a particular cost category?

File: data-pipeline/src/pipeline/config.py [hunk 12]
```diff
@@ -1,45 +1,47 @@
 income_category_map = {
     "academies": {
-        "DFE/EFA Revenue grants (includes Coronavirus Government Funding": "Income_Other DFE grants",
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/config.py:3]: We will need this for the historic incomes though

File: data-pipeline/src/pipeline/pre_processing.py [hunk 13]
```diff
@@ -256,22 +256,150 @@ def prepare_ks4_data(ks4_path):
...
+        ] = 0.0
+
+        if (current_year <= 2022) and (
+            "BNCHBAI061 (Coronavirus Govt Funding)"
+            in central_services_financial.columns
+        ):
+            central_services_financial[
+                "BNCH11123-BTI011-A (MAT Central services - Income)"
+            ] = central_services_financial["BNCHBAI061 (Coronavirus Govt Funding)"]
+
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/pre_processing.py:281]: I get what is happening here, but as discussed, these are not actually substitutions but completely separate.

Prior to 2023 COVID funding was considered, so it should be accounted for.


Then, and what I think is completely separate, BNCH11123-BTI011-A was added to the raw data, and added into the mappings. The two are not mutually interchangeable and I am concerned that anyone picking up the code at a later date may think they are if we do not put a clear difference in between the two.

File: data-pipeline/src/pipeline/main.py [hunk 14]
```diff
@@ -111,8 +111,8 @@ def pre_process_ks4(run_type, year) -> pd.DataFrame:
 
 def pre_process_academy_ar(run_type, year) -> tuple[pd.DataFrame, pd.DataFrame]:
     logger.info("Processing Academy AR Data")
-    academy_ar_data = get_blob(raw_container, f"{run_type}/{year}/academy_ar.xlsx")
-    aar = prepare_aar_data(academy_ar_data)
+    academy_ar_data = get_blob(raw_container, f"{run_type}/{year}/aar.csv")
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/main.py:114]: Looks like this doesn't match the name of the files in `raw`, same with the `aar_cs.csv` file as well.

Comment (Chris-Lorch) [reply data-pipeline/src/pipeline/main.py:114]: Note: In the list of required files, I'll keep the names as aar and aar_cs, so we should just change the name of the file we have in blob store to resolve this.

File: data-pipeline/src/pipeline/input_schemas.py (hunk 2 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/input_schemas.py:315]: Don't know what the T prefix stands for. I was just refering to the cost mapping sheet. This is one for Alan and Assad

File: data-pipeline/src/pipeline/input_schemas.py (hunk 4 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/input_schemas.py:276]: Don't know what the T prefix stands for. I was just refering to the cost mapping sheet. This is one for Alan and Assad

File: data-pipeline/src/pipeline/input_schemas.py (hunk 6 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/input_schemas.py:342]: Don't know what the T prefix stands for. I was just refering to the cost mapping sheet. This is one for Alan and Assad

File: data-pipeline/src/pipeline/input_schemas.py (hunk 7 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/input_schemas.py:344]: Don't know what the T prefix stands for. I was just refering to the cost mapping sheet. This is one for Alan and Assad

File: data-pipeline/src/pipeline/config.py (hunk 12 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/config.py:3]: Unless I'm missing something the conditonal mapping in the `process_aar_data` should handle this.

File: data-pipeline/src/pipeline/pre_processing.py [hunk 15]
```diff
@@ -256,22 +256,150 @@ def prepare_ks4_data(ks4_path):
...
+    if (current_year < 2023) or (
+        "BNCH11123-BTI011-A (MAT Central services - Income)"
+        not in central_services_financial.columns
+    ):
+        central_services_financial[
+            "BNCH11123-BTI011-A (MAT Central services - Income)"
+        ] = 0.0
+
+        if (current_year <= 2022) and (
```
[the remark this answers is not in the document]
Comment (colinbull) [reply data-pipeline/src/pipeline/pre_processing.py:274]: I'll double check this..

File: data-pipeline/src/pipeline/pre_processing.py (hunk 13 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/pre_processing.py:281]: What should be the logic? Is this fundamentally correct?

File: data-pipeline/src/pipeline/mappings.py [hunk 16]
```diff
@@ -164,19 +164,15 @@ def map_maintained_school_status(
 
 
 def map_school_website(url: str):
-    if url is None:
+    if url is None or url == "":
```
Comment (PsypherPunk) [inline data-pipeline/src/pipeline/mappings.py:167]: nitpick: if this is possible, it should be `url: str | None` above.

File: data-pipeline/src/pipeline/mappings.py [hunk 17]
```diff
@@ -164,19 +164,15 @@ def map_maintained_school_status(
...
     return "https://" + url
 
 
-def map_is_pfi_school(pfi: str):
-    return (
-        "PFI School"
-        if pfi is not None and pfi.lower() == "part of pfi"
-        else "Non-PFI school"
-    )
+def map_is_pfi_school(pfi: float):
```
Comment (PsypherPunk) [inline data-pipeline/src/pipeline/mappings.py:174]: nitpick: if this is possible, it should be `url: float | None` above.

File: data-pipeline/src/pipeline/pre_processing.py [hunk 18]
```diff
@@ -57,7 +57,7 @@ def prepare_census_data(workforce_census_path, pupil_census_path):
 
     school_pupil_census = pd.read_csv(
         pupil_census_path,
-        encoding="utf8",
+        encoding='cp1252',
         index_col=input_schemas.pupil_census_index_col,
```
Comment (PsypherPunk) [inline data-pipeline/src/pipeline/pre_processing.py:61]: nitpick: needs a `make lint`…?

**Review by PsypherPunk — APPROVED** [review]
Couple of completely minor typing/formatting points but LGTM.

**Review by jrabbott — APPROVED** [review]
LGTM

File: data-pipeline/src/pipeline/input_schemas.py [hunk 19]
```diff
@@ -262,146 +261,123 @@
...
+    "BNCH21102 (Supply teaching staff - extra note in guidance)": "float",
+    "BNCH21103 (Education support staff)": "float",
+    "BNCH21604 (Educational Consultancy)": "float",
+    "BNCH21606 (Agency supply teaching staff)": "float",
+    "BNCH21403 (Energy)": "float",
+    "BNCH21402 (Water and sewerage)": "float"
 }
 
-aar_central_services_index_col = "Lead UPIN"
+aar_central_services_index_col = "Lead_UPIN"
```
Comment (Chris-Lorch) [inline data-pipeline/src/pipeline/input_schemas.py:326]: Just from our previous convorsations, I thought we were leaving the aar / aar_cs files out until after the private beta release? If we are bringing them in, can we confirm that they do run for all years with the aar / aar_cs data I sent you?

File: data-pipeline/src/pipeline/input_schemas.py (hunk 5 above)
Comment (Chris-Lorch) [reply data-pipeline/src/pipeline/input_schemas.py:328]: @colinbull was there a response to this?

File: data-pipeline/src/pipeline/mappings.py (hunk 8 above)
Comment (Chris-Lorch) [reply data-pipeline/src/pipeline/mappings.py:310]: @colinbull was there a reason for these LAEstabs?

File: data-pipeline/tests/unit/pre_processing/test_aar.py (hunk 9 above)
Comment (Chris-Lorch) [reply data-pipeline/tests/unit/pre_processing/test_aar.py]: @colinbull Are we still using the trust UPIN in the BFR process?

File: data-pipeline/src/pipeline/pre_processing.py (hunk 18 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/pre_processing.py:61]: Yep, not ran make lint on this yet... it is a big enough PR as is :)

File: data-pipeline/workbooks/pre-processing.ipynb [hunk 20]
```diff
@@ -368,8 +368,8 @@
    },
    "outputs": [],
    "source": [
-    "academy_ar = pre_processing.prepare_aar_data('data/academy_ar.xlsx')\n",
-    "central_services = pre_processing.prepare_central_services_data('data/academy_ar.xlsx')"
+    "academy_ar = pre_processing.prepare_aar_data('data/AR7_BenchmarkReport_2022.csv')\n",
```
Comment (Chris-Lorch) [inline data-pipeline/workbooks/pre-processing.ipynb:371]: This should actually be named aar.csv and aar_cs.csv.

File: data-pipeline/src/pipeline/input_schemas.py (hunk 19 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/input_schemas.py:326]: We were but this has now been prioritised in. I have ran them locally for all years 2021, 2022, 2023

File: data-pipeline/src/pipeline/mappings.py (hunk 17 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/mappings.py:174]: Will sort in a following merge

File: data-pipeline/src/pipeline/mappings.py (hunk 8 above)
Comment (colinbull) [reply data-pipeline/src/pipeline/mappings.py:310]: They are taken from the SFB script that generates the London weighting column. I think the specific inclusion / exclusion is to do with bourgh changes.

**Review by Chris-Lorch — APPROVED** [review]

**Pull request merged**